README.txt
# JS8Spotter MIT License, Joe Lyman KF7MIX
# You should have received a copy of the source and license with the program
# Special thanks to KE0DHO, KF0HHR, N0GES, N6CYB, KQ4DRG, KI6ESH, and N4FWD.

# Overview

A JS8Call API mechanism to track activity containing specific search
terms, including callsigns or other activity. Matches on RX.ACTIVITY,
RX.DIRECTED, and RX.SPOT only. Tested with JS8Call v2.2.0


# Settings

JS8Spottter requires the TCP API in JS8Call. To turn it on, go to JS8Call, enter
the "File>Settings" menu, then go to the "Reporting" tab, and enable the
following three checkmarks:

* Allow setting station information
* Enable TCP Server API
* Accept TCP Requests

If you change port or IP from the default, TCP address and port settings for
JS8Spotter may be modified in the application file menu.

Make sure to set your callsign and grid in the application file menu when
running the software for the first time.


# Usage

JS8Spotter is designed to help you keep track of specific, desired traffic as it
passes through JS8Call. From within JS8Call itself, you can access all traffic
via the ALL.TXT log file. JS8Spotter simply provides a method to sort out some
of this traffic as it comes in, and view the traffic in a quick and organized
way.

## Starting the Program

Make sure to start JS8Spotter only after running the JS8Call application, as it
needs to connect to JS8Call in order to function properly. Starting JS8Spotter
first may result in undesired behavior.

## Profiles

A profile in JS8Call is a container where you keep a collection of specific
search terms.

   Example 1: You run a net for your stamp collecting group, with the JS8Call
              group name @STAMPS. You create a profile named "@STAMPS", and
              add search terms for each callsign in the roster, a keyword for
              the group name "@STAMPS", plus several keywords relating to the
              group.

   Example 2: You have a group of close friends who you see on the air often.
              You create a profile named "Contacts" and populate it with search
              terms for each of their callsigns.

   Example 3: You participate in @JS8CHESS. You create a "Chess" profile, and
              populate it with the search term "CHESS" to capture all traffic
              that mentions chess (including anything with "@JS8CHESS". You
              also include search terms for each of the callsigns you are
              playing games with.

JS8Spotter's database file comes with a single profile named "Default". You may
rename this profile to whatever you wish. It may be deleted only after adding
additional profiles.

## Search Terms

As JS8Call runs, it provides decoded activity on the TCP network port.
JS8Spotter watches that traffic, and looks for keywords / search terms that you
provide in each of your profiles. When it finds a match, it saves that traffic
to your specific log (and to a database on the back-end.) The latest 100 entries
are visible at any time, older entries are also visible in the export window
or in the database.

Search terms can be anything you like. They can be a single letter, a word, a
callsign, a JS8Call command, or multiple words.

   Example 1: You have a "Testing" profile, where you track random things that
              are happening on the waterfall. In it, you create a "MSG" search
              term, which lets you capture and save all MSG traffic as it goes
              by. Later, you can look back at any MSGs that were sent. You add
              ">" to your search terms, so you can capture any relays, even if
              they weren't MSG relays.

   Example 2: You are curious about station info, but don't want to always be
              the one requesting it. You create a "Stations" profile, and add
              the search terms "INFO" and "STATUS".

   Example 3: You're part of an emcomm group, and you know that several of the
              group members use JS8Call. You create a profile just for them, and
              add the full roster of callsigns as individual search terms, so
              you can get a feel for when your fellow group members are on the
              air.

## Background Scanning

If enabled, background scanning allows profiles to scan even when the profile is
not actively selected. If disabled, a profile will only record activity when
selected/active.

Background scanning is disabled by default, and may be enabled with the checkbox
at the top-right of each profile. The background scanning setting is set on a
per-profile basis. Settings are retained when the program is closed.

   Example 1: You have a "Testing" profile, which gathers a lot of data. You
              don't want to gather it all the time, only when you are testing,
              so you uncheck the box for that profile.

   Example 2: You don't ever want to miss any JS8CHESS traffic, so you enable
              background scanning on that profile.

   Example 3: Your @STAMPS net is once a month, and you don't need to track
              it except when the net is active. You leave background scanning
              disabled on that profile.

   Example 4: Your want to always know when your friends are around, so you
              keep your "Contacts" profile active with background scanning on.

## User Interface

### Main Window

* The profile drop-down will let you quickly switch between your profiles
* The "Search Term" and "Matched Activity" panes show the corresponding data
* The "Matched Activity" pane only shows the latest 100 entries
* You may double-click on individual search terms or activity for details
* An entry box at the bottom-left allows you to add new search terms
* The "Remove Selected" button allows you to remove the highlighted search term
* The "Export Log" button brings up the export window
* The "Clear Log" button deletes ALL activity entries for the current profile
* Menu actions hopefully perform as labeled

## Export Log Window

The export log window shows all database entries for activity, for the current
profile. The data is tab-delimited. Highlight and copy what you wish, and
paste it into a spreadsheet for easy access. Or, use the "Copy All" or "Save As"
buttons.

## Export Search Terms Window

Allows you to copy/save your search terms, as a backup or to share with others.

## Import Search Terms Window

Allows you to import previously saved search terms, or enter search terms in
bulk, line by line.

## Hide Heartbeats Menu Item

The "Hide Heartbeats" View menu item allows you to hide/show items in the
Matched Activity pane that contain the words "HB" or "Heartbeat". These items
are still retained in the database, they are simply hidden on the screen when
this option is selected. This option is not profile specific. This setting is
retained when the program is closed.

## Simple Offline Map Window

Provides a simple offline maidenhead grid map of the latest 100 contacts.

## Visualize Waterfall Window

This window lets you visualize waterfall activity on a simulated waterfall.
Activity is gathered as you RX on a band. When you view the visual, you'll be
able to quickly get a feel for the parts of the band that are more or less
busy.

## Expect Window

The Expect subsystem allows you to specific text that your station "expects" to
see, along with a response that your station will send automatically if the
expected text is found. All expect commands are prefixed with "E?". You may
specify which stations or groups may trigger a response, and you may specify
how many times your station will respond to a single expect.

See the in-program help for more details.

## Multiple-Choice Forms (MCForms)

The MCForms subsystem extends JS8Call with some basic forms capabilities.
Naturally, you don't want to send large amounts of information over a mode
like JS8Call. MCForms allows you to maintain forms on both sides of the
communiation, and send only the answers to form questions. Forms are .txt
files located in the forms folder. Their format is as follows:

* First line MUST be the human readable title of the form, then the pipe
  symbol |, and then the form ID. Like this example: Test Form|F!001
* A question line starts with a question mark and space, followed by
  the text of the question, as: ? What is your favorite color?
* All answer lines that follow, up to the next question, will be associated.
* Answer lines start with an at symbol and response character, a space, and
  then the text of the answer, as: @1 Red
* A question may have up to 36 answers (A-Z an 0-9 are valid response chars)
* Blank lines are ignored. Lines that start with anything other than ? or @
  are also ignored.
* It is recommended that comments start with #. The first line of the file
  MUST be as described above, and not a comment.

Details of the included forms formats will be posted on KF7MIX.com in the
JS8Spotter section. Operators are encouraged to contact KF7MIX to post any forms
they will be using on the air, so that those formats may be publicly available.

For more form format details, view the included forms in the forms folder.

Filled out forms are posted to the Expect system. When collected over the air,
received form responses will be in the tools menu under form responses for
easy viewing.

The form responses window has a gateway option. If a URL is entered, your system
will automatically forward received forms to your gateway. A test gateway is
available at kf7mix.com, or program your own gateway to receive forms data.

## APRS Commands

JS8Call supports APRSIS gateway commands. The APRS options in the Tools menu
help you format these commands for easy and quick use. Use them for sending
an SMS text, an email, or reporting your grid location.

# Support

I'm a hobbyist programmer, and I've done very little GUI programming. Keep it
in mind.

The software is free, but I'd be happy to lend a hand with it as needed. Contact
KF7MIX via the information on QRZ.com with questions.


73, Joe KF7MIX


CHANGELOG.txt
# JS8Spotter, MIT License, Joe Lyman KF7MIX
# You should have received a copy of the source and license with this application

# Change Log (latest first)

2023-03-24    v1.05b    Fixed bug with spaces in expect allowed list
                        Renamed the "allowed" input in Expect to match the list view name
                        Adjusted frame/grid to help with text cutoff on some systems
                        Added warning on startup if callsign or grid aren't set
                        Added lines to prevent expect from sending when callsign isn't set
                        Added manual
                        Setup github release (thanks to Dan KI6ESH!)
2023-03-07    v1.04b    Fixed bug in export saveas function
                        Added missing comments to report output
                        Added ability to hide RX.SPOT in view menu
                        Added <return> bindings to some list views
                        Added right-click copy activity info to clipboard
                        Added freq,offset,speed to activity database, updated migration script & initdb
                        Added simulated waterfall for visualization of band activity
                        Added some auto database cleanup
                        Added internet gateway for forms reporting, user configurable optional
                        Improvements in activity detail info
2023-02-23    v1.03b    Added status bar for incoming expect requests and form responses
                        Improved the appearance of the double-click activity view
                        Added ability to load a previously posted form answers from expect system into forms
                        Added missing "Export All" in form responses
                        Added copy/export buttons and right-click copy popup to form response pages
                        Added notes & date to form files
                        Fixed bug in short datecode generator
2023-02-14    v1.02b    Fixed scrollbar display issues
                        Fixed row coloring issues
                        Fixed bug when search string was a number
                        Fixed error when right-clicking inside export dialogs with nothing selected
                        Corrected some window size issues, made more windows resizeable
                        Re-arranged map lookup tables to match actual map layout
                        Added manual send to expect
                        Modified * handling in expect to allow groups+* as from
                        Cleaned up code & comments a bit, reorganized many functions
                        Cleaned up some methods around focus() for a little more consistency
                        Added datetime to expect sent record
                        Preserve expect sent record when updating expect entry
                        Changed expect view to scrollable window
                        Added MCForms subsystem (multiple choice forms) - Includes new database table
2023-01-26    v1.01b    Added APRS commands to tools menu. Enabled ability to
                        resize the main window
2023-01-22    v0.99b    Cleaned up some bugs in the expect system, and
                        some Windows-specific bugs in the maps and other areas
2022-12-19    v0.97b    Added TCP settings and network check
2022-12-15    v0.96b    Improved grid map. Added Europe; Added ability to
                        remove items from grid map; Corrected several bugs
2022-12-09    v0.95b    Added a simple offline maidenhead grid map, and
                        fixed some button alignment in Linux
2022-01-25    v0.93a    Fixed race condition in matching, and error handling,
                        which were both causing the program to hang and not
                        record matches
2022-01-20    v0.92a    Switched from UDP to TCP
2022-01-17    v0.91a    Updated initdb program, fixed theme items
2022-01-14    v0.90a    Bug fix impacting background search
2022-01-14    v0.89a    Added external theme, light and dark, bug fixes
2022-01-13    v0.88a    Added import/export and multi-delete for search terms
2022-01-12    v0.87a    Some new key bindings, save as dialog in export
2022-01-11    v0.85a    First public release


js8spotter_initdb.py
# JS8Spotter InitDB v1.04b - 03/07/2023
# Utility to initialize database
#
# MIT License, Copyright 2023 Joseph D Lyman KF7MIX
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

import sqlite3

conn = sqlite3.connect('js8spotter.db')
c = conn.cursor()

c.execute("""CREATE TABLE setting (
    id    INTEGER PRIMARY KEY AUTOINCREMENT,
    name  TEXT    UNIQUE ON CONFLICT IGNORE,
    value TEXT
)
""")

c.execute("""CREATE TABLE profile (
    id     INTEGER PRIMARY KEY AUTOINCREMENT,
    title  TEXT    UNIQUE ON CONFLICT IGNORE,
    def    BOOLEAN DEFAULT (0),
    bgscan BOOLEAN DEFAULT (0)
)
""")

c.execute("""CREATE TABLE activity (
    id         INTEGER   PRIMARY KEY AUTOINCREMENT,
    profile_id INTEGER,
    type       TEXT,
    value      TEXT,
    dial       TEXT,
    snr        TEXT,
    call       TEXT,
    spotdate   TIMESTAMP,
    freq       TEXT,
    offset     TEXT,
    speed      TEXT
)
""")

c.execute("""CREATE TABLE search (
    id         INTEGER   PRIMARY KEY AUTOINCREMENT,
    profile_id INT,
    keyword    TEXT,
    last_seen  TIMESTAMP
)
""")

c.execute("""CREATE TABLE grid (
    grid_callsign VARCHAR (64) UNIQUE ON CONFLICT REPLACE PRIMARY KEY,
    grid_grid VARCHAR (16),
    grid_dial VARCHAR (64),
    grid_type VARCHAR (64),
    grid_snr VARCHAR (16),
    grid_timestamp TIMESTAMP
)
""")

c.execute("""CREATE TABLE "expect" (
	"expect"	VARCHAR(6) UNIQUE ON CONFLICT REPLACE,
	"reply"	    TEXT,
	"allowed"	TEXT,
	"txlist"	TEXT,
	"txmax"	INTEGER,
	"lm"	TIMESTAMP,
	PRIMARY KEY("expect")
)
""")

c.execute("""CREATE TABLE "forms" (
	"id"	INTEGER UNIQUE,
	"fromcall"	TEXT,
	"tocall"	TEXT,
	"typeid"	TEXT,
	"responses"	TEXT,
	"msgtxt"	TEXT,
	"timesig"	TEXT,
	"lm"	TIMESTAMP,
    "gwtx"      TEXT,
	PRIMARY KEY("id" AUTOINCREMENT)
)
""")

c.execute("""CREATE TABLE signal (
    "id"     INTEGER PRIMARY KEY AUTOINCREMENT,
    "sig_callsign" VARCHAR (64),
    "sig_dial"     TEXT,
    "sig_freq"     TEXT,
    "sig_offset"   TEXT,
    "sig_speed"    TEXT,
    "sig_snr"      TEXT,
    "sig_timestamp" TIMESTAMP
)
""")

conn.commit()

new_val = "Default"
c.execute("INSERT INTO profile(title, def) VALUES ('Default', 1)")
c.execute("INSERT INTO setting (name, value) VALUES ('udp_ip','127.0.0.1'),('udp_port','2242'),('tcp_ip','127.0.0.1'),('tcp_port','2442'),('hide_heartbeat','0'),('dark_theme','0'),('marker_index','0'),('wfband_index','0'),('wftime_index','0'),('callsign','FILL'),('grid','FILL'),('hide_spot','0')")

conn.commit()
conn.close()
